LinkBalMonV2(minPerform) and CCIPWatchlist AutoUpdate Contract

poke1994 opened the pull request:
Adding 2 files. 
1. I added a minPerform() function to the LinkBalMonV2 contract and tested it on Sepolia. The purpose of the minPerform function is to establish a minimum batch size required for the sampleUnderfunded address array. If the array size is less than minPerform, no top-up operation will be performed. This optimization is crucial for deploying the BalanceMonitor contract on Ethereum, which consumes significant gas. 

2. Creation of a CCIPOnRampWatchlistUpdate contract to allow auto update of the watchlist on the LinkBalMonV2 when deployed to fund CCIP On-Ramps

Testing File Link - https://docs.google.com/spreadsheets/d/1zk4J4xQrhSJhIEt7l1bpsG6_jrl6qTQLNa5EZia3JzU/edit?gid=829494086#gid=829494086 Specification File - https://docs.google.com/document/d/1p4naJOlrdPmZ5xsZkZE7FDw1cEQqWcQ3hgguhI21SNQ/edit#heading=h.3t9vh460y192

## Ticket
<!---
  Does this work have a corresponding ticket?

  Please link your Jira ticket by including it in one of the following reference:
    - the PR title
    - branch name
    - commit message
  
  By referencing it, it will let the QA team to know what to watch out for when creating a new release.

  Example:

  [LINK-777](https://smartcontract-it.atlassian.net/browse/LINK-777)
--> 

## Requires Dependencies
<!---
  Does this work depend on other open PRs?

  Please list other PRs that are blocking this PR.

  Example:

  - https://github.com/smartcontractkit/chainlink-common/pull/7777777
-->

## Resolves Dependencies
<!---
  Does this work support other open PRs? 

  Please list other PRs that are waiting for this PR to be merged.

  Example:

  - https://github.com/smartcontractkit/ccip/pull/7777777
-->

poke1994 committed:
Add files via upload

Adding 2 files. 
1. I added a minPerform() function to the LinkBalMonV2 contract and tested it on Sepolia. The purpose of the minPerform function is to establish a minimum batch size required for the sampleUnderfunded address array. If the array size is less than minPerform, no top-up operation will be performed. This optimization is crucial for deploying the BalanceMonitor contract on Ethereum, which consumes significant gas. 

2. Creation of a CCIPOnRampWatchlistUpdate contract to allow auto update of the watchlist on the LinkBalMonV2 when deployed to fund CCIP On-Ramps

Testing File Link - https://docs.google.com/spreadsheets/d/1zk4J4xQrhSJhIEt7l1bpsG6_jrl6qTQLNa5EZia3JzU/edit?gid=829494086#gid=829494086
Specification File - https://docs.google.com/document/d/1p4naJOlrdPmZ5xsZkZE7FDw1cEQqWcQ3hgguhI21SNQ/edit#heading=h.3t9vh460y192

poke1994 committed:
Update LinkAvailableBalanceMonitor.sol

poke1994 committed:
Update LinkAvailableBalanceMonitor.sol

FelixFan1992 commented on contracts/src/v0.8/automation/upkeeps/CCIPAutoWatchlist.sol:
```diff
@@ -0,0 +1,94 @@
+import {LinkAvailableBalanceMonitor} from "../upkeeps/LinkAvailableBalanceMonitor.sol";
+
+
+struct Log {
```
can you import this and ILogAutomation from the existing `ILogAutomation.sol` interface?

FelixFan1992 commented on contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:
```diff
@@ -333,7 +337,8 @@ contract LinkAvailableBalanceMonitor is AccessControl, AutomationCompatibleInter
     address[] memory needsFunding = sampleUnderfundedAddresses();
-    if (needsFunding.length == 0) {
+    uint16 minPerform = s_minPerform;
+    if (needsFunding.length <= minPerform) {
```
nit: it's probably more idiomatic to do `<`

FelixFan1992 commented on contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:
```diff
@@ -333,7 +337,8 @@ contract LinkAvailableBalanceMonitor is AccessControl, AutomationCompatibleInter
   ) external view override whenNotPaused returns (bool upkeepNeeded, bytes memory performData) {
     address[] memory needsFunding = sampleUnderfundedAddresses();
-    if (needsFunding.length == 0) {
+    uint16 minPerform = s_minPerform;
```
no need to create a new variable unless it will be read again in this function

FelixFan1992 commented on contracts/src/v0.8/automation/upkeeps/CCIPAutoWatchlist.sol:
```diff
@@ -0,0 +1,94 @@
+    require(routerAddress != address(0), "Router address not set");
+
+    // Define the event signature for OnRampSet(uint64,address)
+    bytes32 eventSignature = keccak256(abi.encodePacked("OnRampSet(uint64,address)"));
```
since this event sig is known in advance, u can encode it and set it as a constant

FelixFan1992 commented on contracts/src/v0.8/automation/upkeeps/CCIPAutoWatchlist.sol:
```diff
@@ -0,0 +1,94 @@
+    function checkLog(
+    Log calldata log,
+    bytes memory checkData
+) external view override returns (bool upkeepNeeded, bytes memory performData) {
```
you can try to run 
`pnpm prettier:write` to format all the code

FelixFan1992 commented on contracts/src/v0.8/automation/upkeeps/CCIPAutoWatchlist.sol:
```diff
@@ -0,0 +1,94 @@
+
+    event setWatchlistOnMonitor(uint64 indexed _dstChainSelector,address indexed _onRamp);
+
+    LinkAvailableBalanceMonitor public LinkAvailableBalanceMonitorAddress;
```
nit: variable starts with lowercase and it should be called `linkAvailableBalanceMonitor` or `monitor` bc it's not an address anymore

FelixFan1992 commented on contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:
```diff
@@ -389,6 +394,12 @@ contract LinkAvailableBalanceMonitor is AccessControl, AutomationCompatibleInter
 
+  /// @notice Update s_minPerform
+  function setMinPerform(uint16 minPerform) public onlyRole(ADMIN_ROLE) {
+    emit MinPerformSet(s_minPerform, minPerform);
```
nit: emit after the variable is set

FelixFan1992 commented on contracts/src/v0.8/automation/upkeeps/CCIPAutoWatchlist.sol:
```diff
@@ -0,0 +1,94 @@
+    if (log.source == routerAddress && log.topics.length > 0 && log.topics[0] == eventSignature) {
+        // Extract the indexed parameter from the log
+        uint64 destChainSelector = uint64(uint256(log.topics[1])); // cast to uint64
+        address onRamp = address(uint160(uint256(bytes32(log.data)))); // extract from log.data
```
depending on the possible values of `destChainSelector` and `onRamp`, maybe add some sanity checks for these values?

FelixFan1992 commented on contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:
```diff
@@ -389,6 +394,12 @@ contract LinkAvailableBalanceMonitor is AccessControl, AutomationCompatibleInter
   }
 
+  /// @notice Update s_minPerform
+  function setMinPerform(uint16 minPerform) public onlyRole(ADMIN_ROLE) {
```
this can be `external`

poke1994 committed:
Update LinkAvailableBalanceMonitor.sol

Made changed as per Felix's comments

poke1994 commented on contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:
```diff
@@ -389,6 +393,12 @@ contract LinkAvailableBalanceMonitor is AccessControl, AutomationCompatibleInter
+  /// @notice Update s_minPerform
+  function setMinPerform(uint16 minPerform) external onlyRole(ADMIN_ROLE) {
+    s_minPerform = minPerform;
+    emit MinPerformSet(s_minPerform, minPerform);
```
@FelixFan1992 I have changed the order of emission as per previous comment

poke1994 commented on contracts/src/v0.8/automation/upkeeps/LinkAvailableBalanceMonitor.sol:
```diff
@@ -333,7 +337,7 @@ contract LinkAvailableBalanceMonitor is AccessControl, AutomationCompatibleInter
   ) external view override whenNotPaused returns (bool upkeepNeeded, bytes memory performData) {
     address[] memory needsFunding = sampleUnderfundedAddresses();
-    if (needsFunding.length == 0) {
+    if (needsFunding.length <= s_minPerform) {
```
@FelixFan1992 I have changed removed the local variable for comparison and now the length of array is directly comparing with the global variable

<= is necessary if minPerform is set to any number >0
If sampleUnderfunded length < minPerform no topUp should be performed

poke1994 committed:
Update CCIPAutoWatchlist.sol

poke1994 reviewed (commented):
Made changes to CCIPAutoWatchlist contract and LinkAvailableBalanceMonitor contract

poke1994 reviewed (commented):
Made changes to LinkBalanceMonitorContract as suggested

FelixFan1992 commented on contracts/src/v0.8/automation/upkeeps/CCIPAutoWatchlist.sol:
```diff
@@ -6,89 +6,67 @@ import "../../shared/access/ConfirmedOwner.sol";
 import {AutomationCompatibleInterface} from "../interfaces/AutomationCompatibleInterface.sol";
 import {AccessControl} from "../../vendor/openzeppelin-solidity/v4.8.3/contracts/access/AccessControl.sol";
 import {LinkAvailableBalanceMonitor} from "../upkeeps/LinkAvailableBalanceMonitor.sol";
+import "../interfaces/ILogAutomation.sol";
```
let's import specific struct / interface

FelixFan1992 commented on contracts/src/v0.8/automation/upkeeps/CCIPAutoWatchlist.sol:
```diff
@@ -0,0 +1,72 @@
+    linkAvailableBalanceMonitorAddress.addToWatchListOrDecommission(_targetAddress, _dstChainSelector);
+  }
+
+  function checkLog(
```
can you add some comments about this function? like `_needsFunding` in the monitor.

and clean up some commented code?

poke1994 replied:
Added comments for updateWatchlist Function
Cleaned up commented code
Imported specific struct and Interface for ILogAutomation

poke1994 committed:
Update CCIPAutoWatchlist.sol

Address comments to the updateWatchlist function of the code
Imported specific parts of ILogAutomation

RyanRHall reviewed (changes requested):
Sorry, but I can't sign off on these without unit tests! From your testing file it looks like you have a good understanding of what needs coverage, but those test cases need to manifest in code :)

chainchad commented:
Going to close this for now as it seems to have gotten stale. Please re-open when ready.
